code/projects/app/src/app/shared/image-download/image-download.component.ts

```diff
@@ -12,9 +19,20 @@ import { dataUrl } from "../../util/image-data";
 export class ImageDownloadComponent implements OnInit {
   @Input({ required: true }) public image$!: Observable<ImageData>;
 
-  protected path = "";
+  protected data: DownloadData = null;
 
   public ngOnInit(): void {
-    this.image$.subscribe((image) => (this.path = dataUrl(image)));
+    this.image$.subscribe((image) => {
+      crypto.subtle.digest("SHA-1", image.data).then((hash) => {
```

Comment:
**note:** Hashing such large buffers might introduce some time overhead, but I haven't experienced it during testing + it's async, so it shouldn't lag the browser.

Reply:
But `async` does not mean anything if its not I/O. It still bogs the main javascript thread.

(correct me if wrong)